File: Problem 117.hs
import Data.Function.Memoize (memoFix)

tiles :: (Num a, Ord a) => (a -> a) -> a -> a
tiles _ 0 = 1
tiles f n
	| n < 0 = 0
	| otherwise = sum $ map f [n-1, n-2, n-3, n-4]

main = print $ memoFix tiles (50 :: Integer)

File: Problem 117.py
memoize = {0: 1}
def tiles(n):
    if n < 0:
        return 0
    if n not in memoize:
        memoize[n] = tiles(n-1) + tiles(n-2) + tiles(n-3) + tiles(n-4)
    return memoize[n]

print(tiles(50))
